File: go-nethttp/store/store.go
package store

import (
	"context"
	"fmt"
	"log"
	"sync"

	"github.com/keith-cullen/microservice/config"
	"github.com/keith-cullen/microservice/store/ent"
	"github.com/keith-cullen/microservice/store/ent/thing"

	_ "github.com/mattn/go-sqlite3"
)

type Store struct {
	mu     sync.Mutex
	Client *ent.Client
}

const (
	DatabaseDriverName = "sqlite3"
)

func Open() (*Store, error) {
	client, err := ent.Open(DatabaseDriverName, config.Get(config.DatabaseFileKey))
	if err != nil {
		return nil, fmt.Errorf("failed to open store: %w", err)
	}
	ctx := context.Background()
	if err := client.Schema.Create(ctx); err != nil {
		return nil, fmt.Errorf("failed to open store: %w", err)
	}
	log.Print("store open")
	return &Store{
		Client: client,
	}, nil
}

func (store *Store) Close() {
	store.Client.Close()
	log.Print("store closed")
}

func (store *Store) GetThing(ctx context.Context, name string) (string, error) {
	store.mu.Lock()
	defer store.mu.Unlock()
	_, err := store.Client.Thing.
		Query().
		Where(thing.Name(name)).
		Only(ctx)
	if err != nil {
		err = fmt.Errorf("failed to get thing: %w", err)
		log.Print(err)
		return "", err
	}
	log.Printf("got thing: %q", name)
	return name, nil
}

func (store *Store) SetThing(ctx context.Context, name string) error {
	store.mu.Lock()
	defer store.mu.Unlock()
	t, err := store.Client.Thing.
		Query().
		Where(thing.Name(name)).
		Only(ctx)
	switch err.(type) {
	case *ent.NotSingularError:
		err = fmt.Errorf("failed to set thing: %w", err)
		log.Print(err)
	case *ent.NotFoundError:
		err = store.createThing(ctx, name)
	default:
		err = store.updateThing(ctx, t)
	}
	return err
}

// unsafe - store.mu must be locked when this method is called
func (store *Store) createThing(ctx context.Context, name string) error {
	_, err := store.Client.Thing.
		Create().
		SetName(name).
		Save(ctx)
	if err != nil {
		err = fmt.Errorf("failed to create thing: %w", err)
		log.Print(err)
		return err
	}
	log.Printf("created thing: %q", name)
	return nil
}

// unsafe - store.mu must be locked when this method is called
func (store *Store) updateThing(ctx context.Context, t *ent.Thing) error {
	_, err := store.Client.Thing.
		UpdateOne(t).
		SetName(t.Name).
		Save(ctx)
	if err != nil {
		err = fmt.Errorf("failed to update thing: %w", err)
		log.Print(err)
		return err
	}
	log.Printf("updated thing: %q", t.Name)
	return nil
}

File: rust-actix-web/src/store.rs
use rusqlite::{Connection, Result, Error, params};
use std::sync::{Arc, Mutex, MutexGuard};
use crate::config;

#[derive(Clone)]
pub struct Store {
    conn: Arc<Mutex<Connection>>,
}

#[derive(Debug)]
struct Thing {
    id: i32,
    name: String,
}

impl Store {
    pub fn new() -> Result<Store> {
        let db_name = config::get(config::DATABASE_FILE_KEY);
        let c = Connection::open(db_name)?;
        c.execute(
            "CREATE TABLE IF NOT EXISTS things (
                        id INTEGER PRIMARY KEY,
                        name TEXT NOT NULL
                )",
            [],
        )?;
        Ok(Store { conn: Arc::new(Mutex::new(c)) } )
    }

    pub fn get_thing(&self, name: &str) -> Option<String> {
        let conn = self.conn.lock().unwrap();
        let op = match self.query_thing(&conn, name) {
            Ok(op) => op,
            Err(e) => {
                println!("Error: Failed to get thing: {}", e);
                return None;
            }
        };
        let thing = match op {
            Some(thing) => thing,
            None => {
                // No entries found or mutliple entries found
                println!("Error: Failed to get thing");
                return None;
            }
        };
        println!("Got thing: \"{}\"", thing.name);
        Some(thing.name)
    }

    pub fn set_thing(&self, name: &str) -> bool {
        let conn = self.conn.lock().unwrap();
        let op = match self.query_thing(&conn, name) {
            Ok(op) => op,
            Err(e) => {
                println!("Error: Failed to set thing: {}", e);
                return false;
            }
        };
        match op {
            None => {
                return self.create_thing(&conn, name);
            }
            Some(thing) => {
                return self.update_thing(&conn, name, thing.id);
            }
        };
    }

    fn query_thing(&self, conn: &MutexGuard<Connection>, name: &str) -> Result<Option<Thing>> {
        let mut stmt = conn.prepare("SELECT id, name FROM things")?;
        let rows = stmt.query_map([], |row| {
            Ok(Thing {
                id: row.get(0)?,
                name: row.get(1)?,
            })
        })?;
        let mut count = 0;
        let mut res = Thing { id: 0, name: "".to_string() };
        for row in rows {
            let thing = row?;
            if name == thing.name {
                res = thing;
                count = count + 1;
            }
        }
        if count < 1 {
            return Ok(None);

        } else if count > 1 {
            return Err(Error::ModuleError("Thing not singular".to_string()));
        }
        Ok(Some(res))
    }

    pub fn create_thing(&self, conn: &MutexGuard<Connection>, name: &str) -> bool {
        if let Err(e) = conn.execute(
            "INSERT INTO things (name) VALUES (?1)",
            params![name],
        ) {
            println!("Error: Failed to create thing: {}", e);
            return false;
        }
        println!("Created thing: \"{}\"", name);
        true
    }

    pub fn update_thing(&self, conn: &MutexGuard<Connection>, name: &str, id: i32) -> bool {
        if let Err(e) = conn.execute(
            "UPDATE things SET name = ?1 WHERE ID = ?2",
            params![name, id],
        ) {
            println!("Error: Failed to update thing: {}", e);
            return false;
        }
        println!("Updated thing: \"{}\"", name);
        true
    }
}
